src/firmware/host/types.rs

```diff
@@ -0,0 +1,492 @@
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+pub struct SnpBuild {
+    /// The version information.
+    pub version: Version,
+
+    /// The build ID.
+    pub build: u32,
+}
+
+/// The platform state.
+///
+/// The underlying SEV-SNP platform behaves like a state machine and can
+/// only perform certain actions while it is in certain states.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[repr(u8)]
+pub enum State {
+    /// The platform is uninitialized.
+    Uninitialized,
+
+    /// The platform is initialized, but not currently managing any
+    /// guests.
+    Initialized,
+
+    /// The platform is initialized and is overseeing execution
+    /// of encrypted guests.
+    Working,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self::Uninitialized
+    }
+}
+
+/// Query the SEV-SNP platform status.
+///
+/// (Chapter 8.3; Table 38)
+#[derive(Default, Debug)]
+#[repr(C)]
+pub struct SnpPlatformStatus {
+    /// The firmware API version (major.minor)
+    pub version: Version,
+
+    /// The platform state.
+    pub state: u8,
+
+    /// IsRmpInitiailzied
+    pub is_rmp_init: u8,
+
+    /// The platform build ID.
+    pub build_id: u32,
+
+    /// MaskChipId
+    pub mask_chip_id: u32,
+
+    /// The number of valid guests maintained by the SEV-SNP firmware.
+    pub guest_count: u32,
+
+    /// Installed TCB version.
+    pub platform_tcb_version: TcbVersion,
+
+    /// Reported TCB version.
+    pub reported_tcb_version: TcbVersion,
+}
+
+/// Rust-friendly instance of the SNP Extended Configuration.
+/// It may be used either to fetch or set the configuration.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct SnpExtConfig {
+    /// SET:
+    ///     Address of the SnpConfig or 0 when reported_tcb does not need
+    ///     to be updated.
+    ///
+    /// GET:
+    ///     Address of the SnpConfig or 0 when reported_tcb should not be
+    ///     fetched.
+    pub config: Option<SnpConfig>,
+
+    /// SET:
+    ///     Address of extended guest request certificate chain or None when
+    ///     previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    ///
+    /// GET:
+    ///     Address of extended guest request certificate chain or None when
+    ///     certificate should not be fetched.
+    pub certs: Option<Vec<CertTableEntry>>,
+
+    /// SET:
+    ///     Length of the certificates.
+    ///
+    /// GET:
+    ///     Length of the buffer which will hold the fetched certificates.
+    pub certs_len: u32,
+}
+
+/// Used to round certificate buffers to 4K page alignment.
+fn round_to_whole_pages(size: usize) -> usize {
+    match size % _4K_PAGE {
+        0 => size,
+        rem => size + (_4K_PAGE - rem),
+    }
+}
+
+impl SnpExtConfig {
+    /// Used to only update the AMD Secure Processor certificates with the certificates provided.
+    pub fn update_certs_only(certificates: Vec<CertTableEntry>) -> Self {
```

Comment:
```suggestion
    pub fn new_certs_only(certificates: Vec<CertTableEntry>) -> Self {
```

Do you prefer this naming?